etcdserver/apply.go

```diff
@@ -745,3 +745,10 @@ func compareInt64(a, b int64) int {
 func isGteRange(rangeEnd []byte) bool {
 	return len(rangeEnd) == 1 && rangeEnd[0] == 0
 }
+
+func noSideEffect(r *pb.InternalRaftRequest) bool {
+	if r.Txn != nil {
+		return isTxnReadonly(r.Txn)
```

Comment:
No. When we start to apply the txn here:

https://github.com/coreos/etcd/blob/master/etcdserver/apply.go#L366
https://github.com/coreos/etcd/blob/master/etcdserver/apply.go#L453

We can actually skip range when there is no waiter. So even if txn is not readonly, we can still skip applying the readonly part of it.